pysimt/translators/beam.py

```diff
@@ -0,0 +1,241 @@
+import logging
+import math
+
+import torch
+
+from ..utils.device import DEVICE
+from ..utils.io import progress_bar
+from ..utils.data import sort_predictions
+from ..models import SimultaneousTFNMT
+
+logger = logging.getLogger('pysimt')
+
+
+"""
+Batched vanilla beam search without any simultaneous translation features.
+"""
+class BeamSearch:
+    def __init__(self, model, data_loader, out_prefix, batch_size, filter_chain=None,
+                 max_len=100, beam_size=5, lp_alpha=0., suppress_unk=False, n_best=False, **kwargs):
+        self.model = model
+        self.data_loader = data_loader
+        self.batch_size = batch_size
+        self.filter_chain = filter_chain
+        self.out_prefix = out_prefix
+        self.beam_size = beam_size
+        self.lp_alpha = lp_alpha
+        self.suppress_unk = suppress_unk
+        self.n_best = n_best
+
+        self.vocab = self.model.trg_vocab
+        self.n_vocab = len(self.vocab)
+        self.unk = self.vocab['<unk>']
+        self.eos = self.vocab['<eos>']
+        self.bos = self.vocab['<bos>']
+        self.pad = self.vocab['<pad>']
+
+        self.max_len = max_len
+        self.do_dump = out_prefix != ''
+        
+        self.is_transformer_model = isinstance(self.model, SimultaneousTFNMT)
+
+    def dump_results(self, hyps, suffix=''):
+        suffix = 'beam' if not suffix else f'{suffix}.beam'
+        suffix += str(self.beam_size)
+
+        # Dump raw ones (BPE/SPM etc.)
+        self.dump_lines(hyps, suffix + '.raw')
+        if self.filter_chain is not None:
+            self.dump_lines(self.filter_chain.apply(hyps), suffix)
+
+    def dump_lines(self, lines, suffix):
+        fname = f'{self.out_prefix}.{suffix}'
+        with open(fname, 'w') as f:
+            for line in lines:
+                f.write(f'{line}\n')
+
+    def decoder_step(self, state_dict, next_word_idxs, h, hypothesis=None):
+        logp, h = self.model.dec.f_next(
+            state_dict, self.model.dec.get_emb(next_word_idxs), h, hypothesis)
+
+        # Similar to the logic in fairseq https://bit.ly/3agXAa7
+        # Never select the pad token or the bos token
+        logp[:, self.pad] = -math.inf
+        logp[:, self.bos] = -math.inf
+
+        return logp, h
+
+    def decoder_init(self, state_dict=None):
+        return self.model.dec.f_init(state_dict)
+
+    def run_all(self):
+        return self.run()
+
+    def run(self, **kwargs):
+        def tile_ctx_dict(ctx_dict, idxs):
+            """Returns dict of 3D tensors repeatedly indexed along the sample axis."""
+            # 1st: tensor, 2nd optional mask
+
+            if self.is_transformer_model:
+                # TF
+                return {
+                    k: (t[:, idxs], None if mask is None else mask[idxs, :])
+                    for k, (t, mask) in ctx_dict.items()
+                }
+            else:
+                # RNN
```

Comment:
is this due to two models having the time axis in different places?

Reply:
Yes, the batch_size is on the first dimension for Transformer encoder mask while for RNN encoder mask the batch_size is on the second dimension.